### PR TITLE
Damage: Play no damage sound when immortal

### LINKED ISSUE #8214: Player damage sound is played when falling from a cliff w/ damage disabled
##### Issue type
 - Bug report
 
##### Minetest version
```
a809f7361e8a7232b3c9ec241be5adcae96b3b2d
```
##### Summary
If you fall from a cliff and drop on a node, you alway hear the player damage sound (`player_damage.ogg`), even if damage is actually disabled.

##### Steps to reproduce
1. Disable damage
2. Start Minetest Game (which contains `player_damage.ogg` but not `player_falling_damage.ogg`)
3. Turn on sound
4. Fall from a cliff on a solid node

### PR DESCRIPTION
**Pull request by SmallJoker**
Add isImmortal server-side for proper enable_damage handling
Rework log messages

Fixes #8214 
Partially implements #8266 

**How to test:**
1) Disable damage of a world
2) Join the world
3) Fall down
4) No damage sound

New log format:
```
2019-03-10 10:26:33: ACTION[Server]: Somebody damaged by 3 hp at (29.262,5.931,-35.038)
2019-03-10 10:26:34: ACTION[Server]: player Somebody (id=3, hp=6) punched player OnceToldMe (id=1, hp=17), damage=1
2019-03-10 10:26:35: ACTION[Server]: player Somebody (id=3, hp=6) punched player OnceToldMe (id=1, hp=17), damage=0
2019-03-10 10:26:35: ACTION[Server]: player Somebody (id=3, hp=6) punched player OnceToldMe (id=1, hp=17), damage=0
2019-03-10 10:26:38: ACTION[Server]: player OnceToldMe (id=1, hp=17) punched player Somebody (id=3, hp=1), damage=5
2019-03-10 10:26:40: ACTION[Server]: player OnceToldMe (id=1, hp=17) punched player Somebody (id=3, hp=0), damage=1
2019-03-10 10:26:40: ACTION[Server]: Somebody dies at (29,4,-35). No bones placed
2019-03-10 10:26:44: ACTION[Server]: Somebody respawns at (15,4,-14.5)
```

### CODE REVIEW

File: src/content_sao.cpp [hunk 1]
```diff
@@ -671,6 +661,16 @@ int LuaEntitySAO::punch(v3f dir,
 		m_env->getScriptIface()->luaentity_on_death(m_id, puncher);
 	}
 
+	char log[256];
+	std::string desc_puncher = puncher->getDescription(),
```
Comment (nerzhul) [inline src/content_sao.cpp]: const std::string &

File: src/content_sao.cpp [hunk 2]
```diff
@@ -671,6 +661,16 @@ int LuaEntitySAO::punch(v3f dir,
...
+			desc_self = getDescription();
+	porting::mt_snprintf(log, sizeof(log),
```
Comment (nerzhul) [inline src/content_sao.cpp]: why not just stream instead of using snprintf here ?

Comment (SmallJoker) [reply src/content_sao.cpp]: IMO, snprintf results in a formatting which is much easier to understand. Or would you prefer a mess like this?
`actionstream << desc_puncher << " (id=" << puncher->getId() << ", hp=" << puncher->getHP() << ") punched " << desc_self << " (id=" << m_id << ", hp=" << m_hp << "), damage=" << (old_hp - (s32)getHP()) << damage_handled ? " (handled by Lua)" : "") << std::endl;`

File: src/content_sao.cpp [hunk 3]
```diff
@@ -628,17 +628,15 @@ int LuaEntitySAO::punch(v3f dir,
 		return 0;
 	}
 
-	ItemStack *punchitem = NULL;
-	ItemStack punchitem_static;
-	if (puncher) {
-		punchitem_static = puncher->getWieldedItem();
-		punchitem = &punchitem_static;
-	}
+	assert(puncher);
```
Comment (nerzhul) [inline src/content_sao.cpp]: we use FATAL_ERROR instead of asserting

File: src/content_sao.cpp (hunk 2 above)
Comment (nerzhul) [reply src/content_sao.cpp]: it's the correct way, you want truncate the log here, description can be very high in terms of characters, same for puncher description. If it's a problem we should consider using a logging library like log4cpp or log4cplus which is more configurable and permits a such pattern without problems

File: src/content_sao.cpp [hunk 4]
```diff
@@ -1268,6 +1271,8 @@ int PlayerSAO::punch(v3f dir,
 	if (!toolcap)
 		return 0;
 
+	assert(puncher);
```
Comment (nerzhul) [inline src/content_sao.cpp]: same here

**Comment by SmallJoker** [comment]
@nerzhul Thanks for your inputs, I applied your requests now.

**Comment by rubenwardy** [comment]
What's the plan with this with respect to #8214

**Comment by SmallJoker** [comment]
@rubenwardy `damage_enabled = false` now sets the armour group to `immortal`. The client already checks for that in `clientenvironment.cpp` when dealing with the collision results and skips the damage audiovisuals.

**Comment by paramat** [comment]
@nerzhul are you able to approve?

**Pull request merged**